Joseph Caguioa

Spring 2020

DS 7337: Natural Language Processing

Section 404 (Tuesday 2030-2200)

HW6 Due: Date of Live Session 11 (3/17/20)

---

# Homework 6

## <u><a name="toc">Table of Contents:</a></u>
* [Question 1](#question1)
* [Question 2](#question2)

---

### <a name="question1">Question 1</a> 

<b>Evaluate text similarity of Amazon book search results by doing the following: 
  
* Do a book search on Amazon. Manually copy the full book title (including subtitle) of each of the top 24 books listed in the first two pages of search results.
* In Python, run one of the text-similarity measures covered in this course, e.g., cosine similarity. Compare each of the book titles, pairwise, to every other one.
* Which two titles are the most similar to each other? Which are the most dissimilar? Where do they rank, among the first 24 results?</b> <sub>[(back to top)](#toc)</sub>

During these days of quarantine, the coronavirus COVID-19 is on everyone's mind. Thus, my Amazon book search topic is on something entirely different for a fun juxtaposition: the sun. In other words, the star at the center of our solar system that many people will not see for themselves until early April, and which happens to have an outer atmosphere called a corona.

In [1]:
from IPython.display import Image
from IPython.core.display import HTML 
Image(url= "https://spaceplace.nasa.gov/sun-corona/en/corona1.en.jpg", 
      width=150, height=150)

The format is first specified as paperback; even with the advent of audiobooks and Kindle-enabled virtual books, physical copies rule. Among the top 24 results, left as sorted by "Featured," are several marked as bestsellers. Results noted as sponsored ads, duplicates, or entries that are not single books (e.g., a trilogy of several books packaged together) are ignored.

In [2]:
titles = ['The Sun Is Also a Star',
          'Embrace the Sun',
          'The Sun and Her Flowers',
          'The Sun Does Shine: How I Found Life, Freedom, and Justice',
          'A Thousand Splendid Suns',
          'A Raisin in the Sun',
          'The Sun Chaser',
          'How Dare the Sun Rise: Memoirs of a War Child',
          'The Warmth of Other Suns: The Epic Story of America\'s Great Migration',
          'I\'ll Give You the Sun',
          'The Sun Also Rises: The Hemingway Library Edition',
          'Dear Sunshine,: A Children\'s Story About The Positive Impact Of The Sun',
          'The Weight of Stars and Suns',
          'Tortilla Sun',
          'The Sun: Our Nearest Star (Let\'s-Read-and-Find-Out)',
          'Where The Sun Goes',
          'Besos de sol, abrazos de luna Sun Kisses, Moon Hugs (Spanish Edition) (Spanish and English Edition)',
          'Gene Wolfe\'s The Book of the New Sun: A Chapter Guide',
          'Half of a Yellow Sun',
          'When the Sun Bursts: The Enigma of Schizophrenia',
          'The Medicine Wheel: Earth Astrology',
          'Twin Suns Of Carrola (Starshatter)',
          'The Sun Is a Compass: My 4,000-Mile Journey into the Alaskan Wilds',
          'In the Old Sun']

In [3]:
from sklearn.feature_extraction.text import CountVectorizer
import pandas as pd
import numpy as np

In [4]:
# Following code modified from https://www.machinelearningplus.com/nlp/cosine-similarity/
count_vectorizer = CountVectorizer(stop_words='english')
count_vectorizer = CountVectorizer()
title_matrix = count_vectorizer.fit_transform(titles)

In [5]:
from sklearn.metrics.pairwise import cosine_similarity

In [6]:
title_similarities = cosine_similarity(title_matrix, title_matrix)
print(title_similarities)

[[1.         0.51639778 0.4        0.28284271 0.         0.4472136
  0.51639778 0.2981424  0.2236068  0.4        0.56568542 0.3721042
  0.18257419 0.31622777 0.42426407 0.4472136  0.09534626 0.38729833
  0.2236068  0.42426407 0.2        0.         0.47809144 0.4472136 ]
 [0.51639778 1.         0.51639778 0.36514837 0.         0.57735027
  0.66666667 0.38490018 0.28867513 0.51639778 0.54772256 0.48038446
  0.23570226 0.40824829 0.36514837 0.57735027 0.12309149 0.5
  0.28867513 0.54772256 0.25819889 0.         0.46291005 0.57735027]
 [0.4        0.51639778 1.         0.42426407 0.         0.4472136
  0.51639778 0.2981424  0.2236068  0.4        0.42426407 0.3721042
  0.36514837 0.31622777 0.42426407 0.4472136  0.19069252 0.38729833
  0.2236068  0.42426407 0.2        0.         0.35856858 0.4472136 ]
 [0.28284271 0.36514837 0.42426407 1.         0.         0.31622777
  0.36514837 0.31622777 0.15811388 0.28284271 0.3        0.26311741
  0.25819889 0.2236068  0.3        0.31622777 0.13483997

In [7]:
# Use to find most similar and most dissimilar
# np.where(title_similarities == title_similarities[title_similarities > 0.0].min())[0]
#title_similarities[title_similarities > 0.0].min()

The most similar titles are <em>A Raisin in the Sun</em> and <em>In the Old Sun</em>, the 4th and 24th results to appear; the cosine similarity between them is 0.75. This is due to the exclusion of stopwords. Without those, the two titles share "Sun" in common and only differ by "Raisin" and "Old." Something similar happens for <em>Embrace the Sun</em> and <em>The Sun Chaser</em>, which would only be off by one word if stopwords were included. In general, the shortest titles (or those pared down with stopwords removed) seem to have the highest similarities with each other.

<u>Top Three Similarities</u>
1. <em>A Raisin in the Sun</em> (4) & <em>In the Old Sun</em> (24): 0.750
2. <em>Embrace the Sun</em> (2) & <em>The Sun Chaser</em> (7): 0.667
3. Different combinations between ranks 2, 7, 12, 16, 24: 0.577

The most dissimilar titles are combinations with no words in common, resulting in a cosine similarity of 0.0. While  nearly all titles that appeared contain "sun," there is no differentiation between the singular and plural forms in my implementation. The two next most dissimilar combinations occur with longer titles with few non-stopwords in common, with one title containing Spanish. In these instances, having more unshared content words drives down the cosine similarity. 

<u>Top Three Dissimilarities</u>
1. Various combinations of books not sharing content words: 0.0
2. <em>Besos de sol, abrazos de luna Sun Kisses, Moon Hugs (Spanish Edition) (Spanish and English Edition)</em> (17) & <em>The Sun Is a Compass: My 4,000-Mile Journey into the Alaskan Wilds</em> (23): 0.0570
3. <em>Dear Sunshine,: A Children's Story About The Positive Impact Of The Sun</em> (12) & <em>Besos de sol, abrazos de luna Sun Kisses, Moon Hugs (Spanish Edition) (Spanish and English Edition)</em> (17): 0.0591

Overall, there does not appear to be an obvious relationship between cosine similarity and ranking in the results (e.g., higher similarities being between higher ranked results). Instead, cosine similarity appears more to be a function of the number of non-stopwords compared across titles.

---

### <a name="question2">Question 2</a> 

<b>Now evaluate using a major search engine.

* Enter one of the book titles from question 1a into Google, Bing, or Yahoo!. Copy the capsule of the first organic result and the 20th organic result. Take web results only (i.e., not video results), and skip sponsored results.  
* Run the same text similarity calculation that you used for question 1b on each of these capsules in comparison to the original query (book title).
* Which one has the highest similarity measure?</b> <sub>[(back to top)](#toc)</sub>

A glance at the summary indicates its contents likely have little to do with tortillas as a main idea, but <em>Tortilla Sun</em> is chosen because of its fun title and the wordplay possibilities from results that did not make the front page. (Additionally, it is one of the few titles that is not also the name of a movie.) 

Bing's first organic capsule is a relevant review of the book on the social cataloging website Goodreads. In contrast, the twentieth organic capsule is a not-as-relevant but tasty-sounding recipe for tortilla soup from Food Network.

In [8]:
title14 = "Tortilla Sun"
capsule1 = "Tortilla Sun had me longing to see the Sandia Mountains, feel the warmth of the sun and hear the call of the wind. Bright New Mexico comes as vividly alive as the colors worn by many of its people. This book is recommended for ages 9 to 12, but I think girls up to 14 or 15 may enjoy it too."
capsule20 = "Simmer Trisha Yearwood's Chicken Tortilla Soup recipe, from Trisha's Southern Kitchen on Food Network, with a jar of salsa, fajita seasoning and canned veggies. All Tortilla Soup Recipes Ideas."
capsule_compare = [title14, capsule1, capsule20]

In [9]:
capsule_matrix = count_vectorizer.fit_transform(capsule_compare)
capsule_similarities = cosine_similarity(capsule_matrix, capsule_matrix)
print(capsule_similarities)

[[1.         0.20701967 0.24253563]
 [0.20701967 1.         0.10041929]
 [0.24253563 0.10041929 1.        ]]


Interestingly enough, the tortilla soup recipe has a slightly higher cosine similarity (0.243) than the book review (0.207) to the book title. This is simply because the review has more non-stopwords than the recipe, and that weighs it down when comparing with the two-word book title.

In [10]:
title14_desc = "When twelve-year-old Izzy spends the summer in her Nana's remote New Mexico village, she discovers long-buried secrets that come alive in an enchanted landscape of majestic mountains, whispering winds, and tortilla suns. Infused with the flavor of the southwest and sprinkled with just a pinch of magic, readers are sure to find this heartfelt story as rich and satisfying as Nana's homemade enchiladas."
capsule_compare2 = [title14_desc, capsule1, capsule20]
capsule_matrix2 = count_vectorizer.fit_transform(capsule_compare2)
capsule_similarities2 = cosine_similarity(capsule_matrix2, capsule_matrix2)
print(capsule_similarities2)

[[1.         0.43989914 0.17977937]
 [0.43989914 1.         0.10041929]
 [0.17977937 0.10041929 1.        ]]


For experimental purposes, Amazon's description of <em>Tortilla Sun</em> is compared against the capsules. Considering that the plot summary contains descriptions of the setting and plot, some of which are mirrored by the reviewer, it is expected that the book review capsule has a higher cosine similarity this time (0.440).

This exercise supports the idea that when it comes to search result display rankings, both on retail sites like Amazon and major engines like Bing, more than just the text similarities are considered when determining what is most relevant to display.

---